modules/app_engine/variables.tf

```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "project_id" {
+  description = "The project to enable app engine on."
+}
+
+variable "location_id" {
+  description = "The location to serve the app from."
+  default     = ""
+}
+
+variable "auth_domain" {
+  description = "The domain to authenticate users with when using App Engine's User API."
+  default     = ""
+}
+
+variable "serving_status" {
+  description = "The serving status of the app."
+  default     = "SERVING"
+}
+
+variable "feature_settings" {
+  description = "A block of optional settings to configure specific App Engine features."
```

Comment:
If I follow correctly, we're expecting the value of `feature_settings` to be a list because `google_app_engine_application.feature_settings` is a block which can technically be declared as a list of maps. In that case, we should convert this to a map and abstract away the need to insert it in a list. Additionally, we should document the expected structure of this variable.

```suggestion

```

Reply:
Ok. The wording here was pulled straight from the Terraform provider documentation and the source docs are terribly worded. The expectation here is that `feature_settings` is a list of map(s). Right now, that list can only consist of a single map with the single key `split_health_checks` so I assume the list of maps is there to allow for future expansion of feature settings maps.

I think the best way to move forward is to do one of two things: either pass the feature settings list into the `google_app_engine_application` resource without any manipulation at all OR we should turn each feature setting into an input. I don't want to start doing list/map manipulation in the resource declaration because doing it correctly will require looping through the list and Terraform doesn't support this case well at all.

I am leaning more towards abstracting out each feature setting into its own input for ease of use by the end user. Thoughts?

Reply:
Considering this matches the provider docs, I lean towards merging this as is.

Reply:
With respect to documenting the structure, I think it would be acceptable to simply link to the underlying resource documentation.

Let me clarify the other point. The underlying resource argument is described as a block, which would be written like:

```hcl
resource "google_app_engine_application" "main" {
  feature_settings {
    split_health_checks = "true"
  }
}
```

Our variable is expected to be a list of maps, which works when passing block arguments via interpolation, but seems unintuitive for the end user. My proposal is that we convert this to a map so that it can be specified equivalent to the preceding example when invoking the module but converted as necessary within the module:

```hcl
resource "google_app_engine_application "main" {
  feature_settings = ["${var.feature_settings}"]
}
```

Reply:
The problem here is that the provider documentation is wrong. The type that Terraform is looking for in the feature_setting field is a list, not a map. The structure here makes little sense to me since it doesn't seem to mirror the API correctly, but it is what it is and I doubt it gets changed in the provider until v3.0.0 of the provider since mirroring the API correctly requires a breaking change.

https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_app_engine_application.go#L71
https://www.terraform.io/docs/extend/schemas/schema-types.html#aggregate-types

Reply:
@thefirstofthe300 pointed out that the actual type is in fact a list, not a block: https://github.com/terraform-providers/terraform-provider-google/blob/master/google/resource_app_engine_application.go#L71.